### PR TITLE
set: specify re-insertion behavior

### PR DESCRIPTION
**Pull request by lucab**
This commit specifies current re-insertion behavior, which does not
perturb insertion order, making such guarantee explicit in docstrings.
It also introduces a duplicate-insert test to ensure the behavior
won't have regressions.

### CODE REVIEW

**Comment by bluss** [comment]
I'll re-check the ordermap code to be sure, but we should also specify if the new or or old key is stored in the given location. Plain Rust HashMap's `.insert(k, v)` will replace an an existing value with `v` but not use the `k`. It's how that one is, and `HashSet` follows that (using only keys).

**Comment by bluss** [comment]
Checked: yes, that's how it is.

1. Preexisting key-value pair is called k1-v1 (if any)
2. On `map.insert(k2, v2)` where `k2` is equivalent to `k1`:
  a. If no key-value pair k1-v1 existed in the map, k2-v2 are stored after insert in the map. (No values dropped)
  b. If k1-v1 already existed in the map, k1-v2 is stored after insert in the map (k2 and v1 are dropped).

The OrderSet works the same as a map -- every element stored in an OrderSet is like a **key** in a map.

**Comment by lucab** [comment]
Good. So, should I leave it as-is or turn it into an explicit "without re-inserting the value nor altering its insertion order"?

File: src/set.rs [hunk 1]
```diff
@@ -145,10 +146,11 @@ impl<T, S> OrderSet<T, S>
         self.map.reserve(additional);
     }
 
-    /// Insert the value into the map.
+    /// Insert the value into the set.
     ///
-    /// If the value did already exist in the set, returns `false`;
-    /// otherwise, return `true`.
+    /// If the value did already exist in the set, return `false`
+    /// without altering its insertion order; otherwise, insert it and
```
Comment (bluss) [inline src/set.rs]: Please mention that the old value remains and that the new is not used/not inserted.

Maybe just -- *if an equivalent value already existed in the set, it remains and the method returns false.*  This covers both parts

Comment (bluss) [reply src/set.rs]: Ok, maybe mention something about order. It remains in the set, in its place in the order. If you want to.

**Comment by lucab** [comment]
Rebased, PTAL.

**Comment by bluss** [comment]
Thanks a lot!

**Pull request merged**